Modularize the git status to support customization

olets opened the pull request:
Modularizes the Git status section! This is discussed in #375.

Until now Spaceship has been wonderfully flexible _except_ in the opinionated Git prompt. This PR breaks the Git prompt section into subsections analogous to the top-level sections in the full prompt, making it easy for users to customize their Git prompt and to build, share, plug in new Git prompt pieces.

Changes:

- Reworks `git.zsh` to run build the Git section from an array of subsections
- Moves the Git status options to their own file, so that they can be easily used in other user subsections
- Documentation

Example usages:

Default:
```zsh
SPACESHIP_GIT_ORDER=(
  branch
  status
)
```

Spaceship < v3
```zsh
SPACESHIP_GIT_ORDER=(
  branch
  status_oh_my_zsh
)
```

Custom
```zsh
SPACESHIP_GIT_ORDER=(
  my_custom_git_prompt
)
```

salmanulfarzy commented on docs/Options.md:
```diff
@@ -149,6 +149,22 @@ Git section is consists with `git_branch` and `git_status` subsections. It is sh
+
+#### Git subsections
+
+The Git section consists of several subsections. You can specify the order of the subsections using `SPACESHIP_PROMPT_ORDER` option. Use Zsh array syntax to define your own prompt order.
```
`git` subsections using `SPACESHIP_GIT_ORDER` option ?

olets replied:
add "`git`"? 👍

salmanulfarzy commented on docs/Options.md:
```diff
@@ -149,6 +149,22 @@ Git section is consists with `git_branch` and `git_status` subsections. It is sh
+
+The Git section consists of several subsections. You can specify the order of the subsections using `SPACESHIP_PROMPT_ORDER` option. Use Zsh array syntax to define your own prompt order.
+
+The order also defines which sections that Spaceship loads. If you're struggling with slow prompt, you can just omit the sections that you don't use, and they won't be loaded.
```
It's already noted at [`Troubleshooting.md`](https://github.com/denysdovhan/spaceship-prompt/blob/master/docs/Troubleshooting.md#why-is-my-prompt-slow), So I don't think this necessary here.

olets replied:
👍

olets replied:
Addressed in ab40f38

salmanulfarzy commented on sections/git.zsh:
```diff
@@ -10,31 +13,32 @@ SPACESHIP_GIT_SHOW="${SPACESHIP_GIT_SHOW=true}"
+  local git_prompt
 
-  local git_branch="$(spaceship_git_branch)" git_status="$(spaceship_git_status)"
+  for el in ${(Oa)SPACESHIP_GIT_ORDER}; do
```
If I'm not mistaken this gives [revers array order](http://zshwiki.org/home/scripting/paramflags#oo_flags), right ? Would be great with a comment here.

olets replied:
That's right. But now that I look at it again, I'll remove the `(Oa)` and flip line 36 to `$git_prompt$(spaceship_git_${el})`

olets replied:
Addressed in c973316

maximbaz commented:
While you are improving git sections, could you perhaps address [this discovery](https://github.com/denysdovhan/spaceship-prompt/issues/375#issuecomment-366461852) as well and move the options to the `git_branch` subsection where they belong?

> `SPACESHIP_GIT_PREFIX` and `SPACESHIP_GIT_SYMBOL` are really branch subsection options.

maximbaz commented on sections/git.zsh:
```diff
@@ -10,31 +13,32 @@ SPACESHIP_GIT_SHOW="${SPACESHIP_GIT_SHOW=true}"
 
-  local git_branch="$(spaceship_git_branch)" git_status="$(spaceship_git_status)"
+  for el in ${(Oa)SPACESHIP_GIT_ORDER}; do
+    source "$SPACESHIP_ROOT/sections/git_${el}.zsh" # load dependency
```
this will source every time a prompt is refreshed, in contrast the rest of the sections are sourced only once - I'd suggest moving the "source" part to `spaceship.zsh` to the same place where other sections are sourced.

olets replied:
Good catch thanks. Addressed in 2b91ca2. I'm just sourcing all git sections since that's what Spaceship does for everything else. (Maybe not ideal, but that can be discussed in #407)

olets commented on sections/git_status_options.zsh:
```diff
@@ -0,0 +1,22 @@
+SPACESHIP_GIT_STATUS_UNTRACKED="${SPACESHIP_GIT_STATUS_UNTRACKED="?"}"
+SPACESHIP_GIT_STATUS_ADDED="${SPACESHIP_GIT_STATUS_ADDED="+"}"
+SPACESHIP_GIT_STATUS_MODIFIED="${SPACESHIP_GIT_STATUS_MODIFIED="!"}"
+SPACESHIP_GIT_STATUS_RENAMED="${SPACESHIP_GIT_STATUS_RENAMED="»"}"
```
I was actually thinking this would be useful for supporting things like #359 — you'd have something like

```
CUSTOM_GIT_SECTION_ADDED="${CUSTOM_GIT_SECTION_ADDED="$SPACESHIP_GIT_STATUS_ADDED"}"
```

`git_status_options.zsh` would be have the basics, and any section might use all or just some of them. That way your `.zshrc` could just have

```
SPACESHIP_GIT_STATUS_ADDED="added yeah!"
```

instead of

```
SPACESHIP_GIT_STATUS_ADDED="added yeah!"
CUSTOM_GIT_SECTION_ADDED="added yeah!"
```

maximbaz replied:
I see, but you wouldn't use two sections at the same time, would you? You'd choose one of the sections, then look up its available options and what they do, and configure to your liking 🙂

olets replied:
Ha true. My thought was this would make it easier for people to try out other git prompts — they could simply change the line in the git prompt array, without changing a bunch of variable names. It also gives hypothetical git prompt developers a starting place

That said I don't strong feelings about it :) maybe we can get a third opinion

maximbaz replied:
Agreed, my final note would be that different modules could (and probably will) reuse the same variable names anyway, for example in my custom module I did use `SPACESHIP_GIT_STATUS_ADDED` variable name 🙂

olets commented:
@salmanulfarzy @maximbaz I've pushed some commits addressing your feedback

@maximbaz I actually think I was wrong about `SPACESHIP_GIT_PREFIX` and `SPACESHIP_GIT_SYMBOL`

maximbaz commented:
> @maximbaz I actually think I was wrong about SPACESHIP_GIT_PREFIX and SPACESHIP_GIT_SYMBOL

Could you elaborate? Because currently I also think those two belong to the branch section 🙂

olets commented:
@maximbaz it's been long enough that I forgot exactly how it's set up :)

`SPACESHIP_GIT_SYMBOL` _should_ be in the branch subsection!

In this PR, the Git section has a prefix, and the branch subsection has one too, so `SPACESHIP_GIT_PREFIX` is good as is. Unless I'm misreading again :)

maximbaz commented:
I thought we could cleanup the `sections/git.zsh` from all configuration options except `SPACESHIP_GIT_SHOW`, for example by renaming `SPACESHIP_GIT_BRANCH_SYMBOL` and moving it into the `sections/git_branch.zsh`. But if that's not trivial and/or doesn't fit into your changes, better leave out for another PR.

olets commented on sections/git_branch.zsh:
```diff
@@ -8,7 +8,7 @@
 
 SPACESHIP_GIT_BRANCH_SHOW="${SPACESHIP_GIT_BRANCH_SHOW=true}"
-SPACESHIP_GIT_BRANCH_PREFIX="${SPACESHIP_GIT_BRANCH_PREFIX="$SPACESHIP_GIT_SYMBOL"}"
+SPACESHIP_GIT_BRANCH_PREFIX="${SPACESHIP_GIT_BRANCH_PREFIX="$SPACESHIP_GIT_BRANCH_PREFIX"}"
```
Hmm that can't be right! I'm sure I meant to say ` `. Will fix

olets commented:
Looking at this again to see where it stands.

I moved the `git symbol` into the `branch` file, as requested — thanks for sticking to your guns on that idea @maximbaz, you were right.

I personally still like the idea of a `git_status_options.zsh` — that way any contributor has the option of pulling in an existing Spaceship default. In fact, it could make sense to encourage using reusing variables from `git_status_options.zsh`, so that if a user wanted to switch from one git prompt to another they'd see familiar symbols in familiar parts of the prompt. (I didn't go back and read the full inline discussion from earlier, but I bet I'm just repeating myself 😉)
